src/Refinery/ProofState.hs

```diff
@@ -185,13 +190,24 @@ proofs s p = go s [] p
           in go s' goals p
       go s goals (Alt p1 p2) = liftA2 (<>) (go s goals p1) (go s goals p2)
       go s goals (Interleave p1 p2) = liftA2 (interleave) (go s goals p1) (go s goals p2)
-      go s goals (Commit p1 p2) = go s goals p1 >>= \case
-          (rights -> []) -> go s goals p2
-          solns -> pure solns
+      go s goals (Commit p1 p2 k) = fmap rights (proofs s p1) >>=
+        \case
+          [] -> fmap rights (proofs s p2) >>= blarg goals k
+          solns -> blarg goals k solns
       go _ _ Empty = pure []
       go _ _ (Failure err _) = pure [throwError err]
       go s goals (Axiom ext) = pure [Right $ Proof ext s goals]
 
+      blarg
+          :: forall a
+           . [goal]
+          -> (a -> ProofStateT ext ext err s m goal)
+          -> [Proof ext s a]
+          -> m [Either err (Proof ext s goal)]
+      blarg goals k solns =
+        fmap join $ for solns $ \(Proof _ s' as) ->
```

Comment:
a bit dodgy that this ignores the extract --- hoping you can weigh in on whether that's important

Reply:
maybe the continuation just needs to take both. that's probably the right answer

Reply:
i've convinced myself that that's the right answer